Fix tracking order details events multiple times

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #4779 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR makes `OrderDetailRepository` to be injected in `ViewModelScope` and moves analytics tracking from `Repository` to `ViewModel` for Order Details screen. 

It also unregisters `OrderDetailRepository` from EventBus dispatcher, allowing garbage collector to collect those objects.

This will assert that:
1. for every object that will use `OrderDetailRepository` in `OrderDetailViewModel` only one instance of repository will be created and released in `OrderDetailViewModel` too (case of `CardReaderPaymentCollectibilityChecker`)
2. For nested screens (like `CreateShippingLabel`) a new instance of repository will be created (so `continuation`s for different ViewModels) will work independently but as we moved sending analytics events to `OrderDetailViewModel` which is always one for details screen - the events will be send only once.

For more detailed explanations, please visit the Slack thread: p1631609358257000/1631557182.228900-slack-C6H8C3G23

This is how it **did** look for "change order status" action.
![image](https://user-images.githubusercontent.com/5845095/133259196-f25294d7-48c4-423c-b4eb-c5731a4795b1.png)

(these logs were added locally, don't expect them to see on your build).


### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Before testing this change you might want to reproduce the issue on your own. Please see #4779 for instructions how to do it.

#### Test sending multiple events fix

1. Open the app
2. Go to list of orders
3. Go to details of any order
4. Go back to list
5. Repeat 3 and 4 for a few times (like 5 times)
6. Go again to detail of any order
7. Change the status of the order
8. **Assert there's exactly one `order_status_change_success` event logged**

#### Test fix for leak of `OrderDetailRepository`

1. Apply PATCH from the bottom of PR description. It adds logs for better visibility
2. Run the app
3. Open logs. You can filter by `PR-4780`
4. Open details of any order. **Assert there's log "Creating instance"**
5. Close details of any order. **Assert there's log "Destructing instance" with the same object reference as above**
6. Perform 4 and 5 several times.

Perform above steps for other screens like `Issue refund` or `Create shipping label`.

---

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt	(revision 93ab9093138feb43ab792a024edb9887c93e1662)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt	(date 1631710984335)
@@ -35,6 +35,7 @@
 import org.wordpress.android.fluxc.store.*
 import org.wordpress.android.fluxc.store.WCOrderStore.*
 import org.wordpress.android.fluxc.store.WCProductStore.OnProductChanged
+import org.wordpress.android.util.AppLog
 import javax.inject.Inject
 
 @ViewModelScoped
@@ -56,9 +57,14 @@
     private var continuationDeleteShipmentTracking = ContinuationWrapper<Boolean>(ORDERS)
 
     init {
+        AppLog.d(AppLog.T.API, "PR-4780 Creating instance: $this")
         dispatcher.register(this)
     }
 
+    protected fun finalize() {
+        AppLog.d(AppLog.T.API, "PR-4780 Destructing instance: $this")
+    }
+
     fun onCleanup() {
         dispatcher.unregister(this)
     }

```





- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Make OrderDetailRepository a singleton

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:
```diff
@@ -10,32 +10,11 @@ import androidx.lifecycle.SavedStateHandle
-import com.woocommerce.android.analytics.AnalyticsTracker.Stat.RECEIPT_PRINT_FAILED
-import com.woocommerce.android.analytics.AnalyticsTracker.Stat.RECEIPT_PRINT_SUCCESS
-import com.woocommerce.android.analytics.AnalyticsTracker.Stat.RECEIPT_PRINT_TAPPED
+import com.woocommerce.android.analytics.AnalyticsTracker.Stat.*
```
I hope those import changes are fine - I've use project's IDE config and I think we agreed to use `*` imports.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt:
```diff
@@ -36,8 +35,9 @@ import org.wordpress.android.fluxc.store.*
 import javax.inject.Inject
+import javax.inject.Singleton
 
-@OpenClassOnDebug
```
I haven't found a usage for that, I think it's safe to remove it

wzieba committed:
Bring back OrderDetailRepository#onCleanup

wzieba committed:
Set scope of OrderDetailRepository to ViewModelScope

wzieba commented:
Following up to our discussion on Slack: we eventually don't have to call `OrderDetailRepository#onCleanup` in `CardReaderPaymentCollectibilityChecker` because after applying `@ViewModelScoped` on `OrderDetailRepository` the same instance of the repository is shared between `OrderDetailViewModel` and `CardReaderPaymentCollectibilityChecker`. The repository has been already cleared inside `OrderDetailViewModel` so there's no need to do this one more time in `CardReaderPaymentCollectibilityChecker`.

WDYT about this? Does it work for you?

wzieba commented:
I leave it as `draft` for now as I know @malinajirka you work on a stateless-repository fix. But in case if some bumps will occur, this PR will fixed our issue so I'd appreciate review of this PR (now or later), just in case. Thanks!

malinajirka commented:
> I leave it as draft for now as I know @malinajirka you work on a stateless-repository fix.

@wzieba I'd vote for reviewing and merging this either way. I'm afk next week and I can't imagine I'd wrap up all the methods and merge all the PRs in time.

wzieba commented:
Ok gotcha - so I'm making it not a draft now then :)

wzieba committed:
Cleanup OrderDetailRepository in every other viewmodel

hichamboushaba commented:
@wzieba just to understand the logic now before reviewing, since the description still mentions changing the class to Singleton 🙂.
Will this fix the multiple track events? I'm asking because since the `Repository` is still not a singleton, I think we may have some cases of duplicate events (less for sure because of the multiple cleanups). An example, let's say for example we open `OrderDetailRepository`, then we open a subscreen that depends on `OrderDetailRepository`, even with the `@ViewModelScope`, we will have two instances of `OrderDetailRepository`, so if the subscreen does something that triggers an event, it will be sent two times.

wzieba commented:
Exactly, you're right @hichamboushaba - I've hurried up with the PR. Let me correct this and then request review. Thank you for noticing, I was thinking about this case when you answered me with moving tracking to ViewModel (here: p1631703846322900/1631557182.228900-slack-C6H8C3G23) but eventually forgot about it 🤦 . Great eye!

wzieba commented:
@hichamboushaba @malinajirka I still struggle with finding a good fix for both issues (leaks and duplicated events).

Because even if we'll move tracking events from Repository to ViewModel, we'll still have duplication: e.g. scenario like this: 
1. Open Order Details screen. This creates and registers Repository (A)
2. Navigate to Issue Refund. `onCleared` of the Order Details screen is not called
3. Issue Refund create and registers Repository (B) (it's already a different ViewModel scope)
4. We change status in Issue Refund (or call any method that will send track analytics) - and here we have Repository A and B registered for the same event hence duplication.

👆This is basically scenario described by @hichamboushaba in comment above, a little more detailed.

---

So I want to propose another idea based on Fragments lifecycles. I'm just about to push a WIP commit - later I'll remove it to not cause confusion. 

The idea is to register Repository in `onStart` and unregister it in `onPause`. Simple as it is, I think it fixes all of our problems. In the WIP commit I've added it only for `OrderDetailViewModel` but eventually I want to add it for every other ViewModel that uses `OrderDetailRepository`.

WDYT? Does it make sense to you? Do you see some better solution?

hichamboushaba commented:
> Because even if we'll move tracking events from Repository to ViewModel, we'll still have duplication: e.g. scenario like this:
> 
> Open Order Details screen. This creates and registers Repository (A)
> Navigate to Issue Refund. onCleared of the Order Details screen is not called
> Issue Refund create and registers Repository (B) (it's already a different ViewModel scope)
> We change status in Issue Refund (or call any method that will send track analytics) - and here we have Repository A and B registered for the same event hence duplication.
> 👆This is basically scenario described by @hichamboushaba in comment above, a little more detailed.


Maybe I'm missing something, but there will be no leak here, since both repositories are still active depending on their corresponding ViewModels, and if we move the track events to the ViewModel, then we can be sure they will be sent only once, since only one continuation is active per time.

> The idea is to register Repository in onStart and unregister it in onPause. Simple as it is, I think it fixes all of our problems. In the WIP commit I've added it only for OrderDetailViewModel but eventually I want to add it for every other ViewModel that uses OrderDetailRepository.

This would fix the issue, but IMHO since we are planning to refactor the repository to get rid of EvenBus completely, it's not worth it (as per my understanding above), and would just make the migration harder, just my 2 cent 🙂

wzieba committed:
Track analytics in OrderDetailViewModel

wzieba commented:
Thank you @hichamboushaba, you're right - I think I finally got the idea. 

For some reason I locked my mind that we will need analytic event for every inner ViewModel separately (Issue Refund, Create Shipping Label etc.). We don't need them though because we'll always have OrderDetailViewModel alive and registered to listen to EventBus events.

@hichamboushaba @malinajirka I've updated PR and description - I think its finally ready for review :)

(with force-push I've removed the previous, "idea only" commit to not mess git history with pure WIP commit)

hichamboushaba commented:
Thanks @wzieba, I'll take a look this afternoon.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt:
```diff
@@ -37,7 +37,7 @@ import org.wordpress.android.fluxc.store.WCOrderStore.*
 import javax.inject.Inject
 
-@OpenClassOnDebug
+@ViewModelScoped
```
Do you think we should add a comment here explaining that this is needed now to fix a leak, but once we get rid of EventBus, we won't need it?

wzieba replied:
Sure, added in 64a16ff3454f7b7c37d0e4e08eb472303bd92230

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -640,11 +619,62 @@ class OrderDetailViewModel @Inject constructor(
+                        Stat.ORDER_STATUS_CHANGE_FAILED,
+                        prepareTracksEventsDetails(event)
+                    )
+                } else {
+                    AnalyticsTracker.track(Stat.ORDER_STATUS_CHANGE_SUCCESS)
+                }
+            }
+            WCOrderAction.POST_ORDER_NOTE -> {
+                if (event.isError) {
+                    AnalyticsTracker.track(
+                        Stat.ORDER_NOTE_ADD_FAILED,
+                        prepareTracksEventsDetails(event)
+                    )
+                } else {
+                    AnalyticsTracker.track(Stat.ORDER_NOTE_ADD_SUCCESS)
+                }
+            }
+            WCOrderAction.ADD_ORDER_SHIPMENT_TRACKING -> {
+                if (event.isError) {
+                    AnalyticsTracker.track(
+                        Stat.ORDER_TRACKING_ADD_FAILED,
+                        prepareTracksEventsDetails(event)
+                    )
+                } else {
+                    AnalyticsTracker.track(Stat.ORDER_TRACKING_ADD_SUCCESS)
+                }
+            }
+            WCOrderAction.DELETE_ORDER_SHIPMENT_TRACKING -> {
+                if (event.isError) {
+                    AnalyticsTracker.track(
+                        Stat.ORDER_TRACKING_DELETE_FAILED,
+                        prepareTracksEventsDetails(event)
+                    )
+                } else {
+                    AnalyticsTracker.track(Stat.ORDER_TRACKING_DELETE_SUCCESS)
+                }
+            }
+            else -> {
+                // no-op
+            }
```
We had different ideas about how to implement this 🙂

When I suggested moving the tracking to the ViewModel, I meant it as a responsibility for the ViewModel, and as a reaction to the `Repository`'s call, but now I see that what I suggested is not possible right now, it would require refactoring the Repository to return a proper `Result` instead of a Boolean, and on failures, we would need a proper Exception that holds what we need for tracks.

In the future, we can think of something like:

1. first change the signature of the Repository functions to something like:
```kotlin
suspend fun updateOrderStatus(orderModel: WCOrderModel, newStatus: String): Result<Unit> {
```
2. Then consume it in the ViewModel like this:
```kotlin
orderDetailRepository.updateOrderStatus(order.toDataModel(), newStatus)
                    .fold(
                        onSuccess = {
                            AnalyticsTracker.track(Stat.ORDER_STATUS_CHANGE_SUCCESS)
                        },
                        onFailure = {
                            val error = (it as? OrderException)?.error
                            AnalyticsTracker.track(
                                Stat.ORDER_STATUS_CHANGE_FAILED,
                                mapOf(
                                    AnalyticsTracker.KEY_ERROR_CONTEXT to this::class.java.simpleName,
                                    AnalyticsTracker.KEY_ERROR_TYPE to error?.type.toString(),
                                    AnalyticsTracker.KEY_ERROR_DESC to error?.message
                                )
                            )
                        }
                    )
```

Where the `OrderException` is a class specific to this repository:
```kotlin
    class OrderException(val error: OrderError) : Exception()
```

wzieba replied:
> In the future, we can think of something like:

Yeah, I think we can eventually have a proper way of doing that after all PRs related to #4791 will be merged. I agree it'd be better to navigate and understand what will happen for a result of any `OrderDetailRepository` call.

hichamboushaba reviewed (approved):
LGTM, the events are being sent only once, thanks for the fix @wzieba.

I left two comments, but only for discussion.

wzieba committed:
Add a comment about ViewModelScoped usage

wzieba commented:
Thank you @hichamboushaba for review!

@malinajirka do you want to take a look at this too?

malinajirka reviewed (approved):
@wzieba I haven't done a full review but the code looks good to me. Great investigation and great job implementing this 🙇 ! Also, nice work catching the issues @hichamboushaba 👏 !

Feel free to merge this ;).
